bin/varnishd/cache/cache_tcp_pool.c

```diff
@@ -279,32 +307,121 @@ VCP_Rel(struct conn_pool *cp)
 		Lck_Unlock(&conn_pools_mtx);
 		return (1);
 	}
-	AZ(cp->n_used);
+	AZ(cp->stats->n_used);
 	VTAILQ_REMOVE(&conn_pools, cp, list);
+	VSC_C_main->vcp_destroy++;
 	Lck_Unlock(&conn_pools_mtx);
 
 	Lck_Lock(&cp->mtx);
 	VTAILQ_FOREACH_SAFE(pfd, &cp->connlist, list, pfd2) {
 		VTAILQ_REMOVE(&cp->connlist, pfd, list);
-		cp->n_conn--;
+		cp->stats->n_conn--;
 		assert(pfd->state == PFD_STATE_AVAIL);
 		pfd->state = PFD_STATE_CLEANUP;
 		(void)shutdown(pfd->fd, SHUT_WR);
 		VTAILQ_INSERT_TAIL(&cp->killlist, pfd, list);
-		cp->n_kill++;
+		cp->stats->n_kill++;
 	}
-	while (cp->n_kill) {
+	while (cp->stats->n_kill) {
 		Lck_Unlock(&cp->mtx);
 		(void)usleep(20000);
 		Lck_Lock(&cp->mtx);
 	}
 	Lck_Unlock(&cp->mtx);
 	Lck_Delete(&cp->mtx);
-	AZ(cp->n_conn);
-	AZ(cp->n_kill);
+	AZ(cp->stats->n_conn);
+	AZ(cp->stats->n_kill);
+	cp->stats = NULL;
+	VSC_vcp_Destroy(&cp->vsc);
+	AZ(cp->vsc);
 	return (0);
 }
 
+/*--------------------------------------------------------------------
+ * Open a new connection from pool.  This is a distinct function since
+ * probing cannot use a recycled connection.
+ */
+
+static int
+VCP_Open(struct conn_pool *cp, double tmo, const void **privp)
+{
+	int r;
+	double h;
+
+	CHECK_OBJ_NOTNULL(cp, CONN_POOL_MAGIC);
+
+	while (cp->holddown > 0) {
+		Lck_Lock(&cp->mtx);
+		if (cp->holddown == 0) {
```

Comment:
This `if` statement is still inconsistent with the `while` above.

Reply:
github had shown me your initial comment on this as outdated, so I thought you had taken it back.

So what is happening here is that holddown can be modified concurrently to a thread in open. We want to avoid taking the lock each time, so we accept to miss a holddown in favor of somehow less locking overhead. But for the reverse case that the holddown has been cleared concurrently, we want to ensure that we actually see it cleared.

We we check unprotected first but then, under the lock, we make sure that holddown has not been cleared concurrently.

Reply:
BTW, the main motivation for the lock is to get a consistent view on holddown and holddown_errno. While we could juggle membars, a simple mutex is easier and safe and with the above optimization the whole thing has close to no overhead for the base case (no connection errors).

Reply:
I cleaned my glasses and finally was able to see that glaring lock in-between. Thank you for your patience :)